File: examples/merchant-starter-node/src/oauth/store.js
/**
 * In-memory authorization code store。
 *
 * Prod 換成 Redis / DB（code 短效 10 分鐘；只用一次後丟）。
 */

const CODE_TTL_MS = 10 * 60 * 1000;
const codes = new Map();

function putCode(code, payload) {
  codes.set(code, { ...payload, expires_at: Date.now() + CODE_TTL_MS });
}

function consumeCode(code) {
  const entry = codes.get(code);
  if (!entry) return null;
  codes.delete(code); // 一次性
  if (Date.now() > entry.expires_at) return null;
  return entry;
}

// 假會員 DB — 真實整合換成自家 user 查詢
const MEMBERS = {
  "alice@acme.com": { id: "mbr_1234", name: "Alice", email: "alice@acme.com" },
  "bob@acme.com": { id: "mbr_5678", name: "Bob", email: "bob@acme.com" },
};

function authenticateMember(email, password) {
  // Demo: 任何 password 皆通過；真實要 hash compare
  const user = MEMBERS[email];
  if (!user || !password) return null;
  return user;
}

module.exports = { putCode, consumeCode, authenticateMember };

File: examples/merchant-starter-python/app/oauth/store.py
"""In-memory authorization code + 假會員 DB。

Prod 換成 Redis（code 短效 10 分鐘；一次性）+ 自家 user table（hash compare）。
"""

from __future__ import annotations

import time
from dataclasses import dataclass

CODE_TTL_SEC = 10 * 60


@dataclass
class CodeEntry:
    client_id: str
    redirect_uri: str
    member_id: str
    scope: str
    code_challenge: str
    code_challenge_method: str
    expires_at: float


_codes: dict[str, CodeEntry] = {}


def put_code(code: str, entry: CodeEntry) -> None:
    _codes[code] = entry


def consume_code(code: str) -> CodeEntry | None:
    entry = _codes.pop(code, None)
    if entry is None or time.time() > entry.expires_at:
        return None
    return entry


_MEMBERS: dict[str, dict] = {
    "alice@acme.com": {"id": "mbr_1234", "name": "Alice", "email": "alice@acme.com"},
    "bob@acme.com": {"id": "mbr_5678", "name": "Bob", "email": "bob@acme.com"},
}


def authenticate_member(email: str, password: str) -> dict | None:
    """Demo: 任何 password 皆通過；prod 必走 hash compare。"""
    if not email or not password:
        return None
    return _MEMBERS.get(email)
